doofinder-for-woocommerce/includes/api/endpoints/class-endpoint-product.php

```diff
@@ -0,0 +1,521 @@
+     *
+     * @param array $id Product ID selected
+     * @return string The image link
+     */
+    public static function get_image_link($id)
+    {
+        $post = get_post($id);
+        $thumbnail = new Thumbnail($post);
+        $image_link = $thumbnail->get();
+        if (empty($image_link) && $post->post_type === 'product_variation') {
+            $thumbnail = new Thumbnail(get_post($post->post_parent));
+            $image_link = $thumbnail->get();
+        }
+
+        //If neither the variant and the product have an image, return the woocommerce placeholder image
+        $image_link = empty($image_link) ? wc_placeholder_img_src(Thumbnail::get_size()) : $image_link;
+        $image_link = self::add_base_url_if_needed($image_link);
+
+        return $image_link;
+    }
+
+    /**
+     * Check that image link is absolute, if not, add the site url
+     *
+     * @param string $image_link
+     * @return string $image_link
+     */
+    private static function add_base_url_if_needed($image_link)
+    {
+        if (0 === strpos($image_link, "/")) {
+            $image_link = get_site_url() . $image_link;
+        }
+        return $image_link;
+    }
+
+    /**
+     * Field names to exchange
+     *
+     * @param array $product The product array to process.
+     * @return array $product without fields excluded
+     */
+    private static function clean_fields($product){
+        $product["title"] = $product["name"];
+        $product["link"] = $product["permalink"];
+
+        unset($product["name"]);
+        unset($product["permalink"]);
+
+        return $product;
+    }
+
+    /**
+     * Check availability product
+     *
+     * @param array $product The product array to process.
+     * @return array $product with availability string type (in stock / out of stock)
+     */
+    private static function check_availability($product){
+        if($product["purchasable"] && ($product["stock_status"] == "instock" || $product["stock_status"] == "onbackorder")){
+            $product["availability"] = "in stock";
+        }
+        else{
+            $product["availability"] = "out of stock";
+        }
+        return $product;
+    }
+
+    /**
+     * Clears image fields from a product array.
+     *
+     * @param array $product The product array to process.
+     * @return array The product array with empty image fields removed.
+     */
+    private static function clear_images_fields($product){
+        $product["image_link"] = self::get_image_link($product["id"]);
+        unset($product["images"]);
+
+        return $product;
+    }
+
+    /**
+     * Cascade variants to their parent products.
+     *
+     * @param array $products The array of product data.
+     * @return array The modified array of product data with variants cascaded.
+     */
+    private static function cascade_variants($products) {
+        foreach ($products as $key => $product) {
```

Comment:
me podrías explicar por qué necesitamos esto? he visto que se llama a la función `get_variants` y con eso obtienes las variantes de cada producto. por qué necesitas después hacer este recorrido de productos?
me refiero por qué en `get_variants` en vez de hacer 
```
$products = array_merge($products, $variations_data);
```
no haces 
```
if($product["parent_id"] == 0){
    $attr_variation_true = self::get_df_variants_information($product);
    $product["df_variants_information"] = $attr_variation_true;
    $product["variants"] = $variations_data;
    $products[]                         = $product;

```

que además tampoco me queda claro por qué hace falta el `if($product["parent_id"] == 0){` porque me da la impresión de que en esa parte del código sólo entras si es el producto padre y estás preguntando por las variantes..

pero igual me estoy perdiendo algo

Reply:
Si, te explico, básicamente hay 2 funciones diferentes: **get_variations() y cascade_variants()**. La primera lo único que hace es traerse todas las variaciones de los productos y añadirlas al array principal de productos, pero luego necesitamos la función de cascade_variants() para "reconstruir" el array, ya que necesitamos que todas las variaciones de productos se inserten bajo el nodo "variants" de cada producto padre. Es decir, las variaciones no deben estar en el nivel de los padres, tienen que estar dentro de cada padre, tal y como antes se hacía en items transformations.

Reply:
sí, gracias por la explicación! Es verdad que items transformation lo hacíamos así, pero revisándolo  no termino de ver claro que sea necesario hacerlo en dos iteraciones diferentes. Si en la parte de get_variations, en la última parte en vez de hacer el merge al mismo nivel, hacer algo así:
```
if($product["parent_id"] == 0){
    $attr_variation_true = self::get_df_variants_information($product);
    $product["df_variants_information"] = $attr_variation_true;
    $product["variants"] = $variations_data;
    $products[]                         = $product;
```
nos ahorraríamos la función de cascade_variants completa, no?

Reply:
Estoy de acuerdo con @sofia-doofinder, quedaría más optimizado procesando todo en un único bucle.

Reply:
He optimizado la función get_variations para que solo tenga 2 niveles de iteración y ha quedado mucho mejor. 

Respecto a lo que comentáis de fusionar las 2 funciones de **get_variants** y **cascade_variants**, he estado intentando alguna alternativa pero es muy complejo ya que esto se ha hecho así porque primero se piden todas las variantes de todos los productos y se dejan en primer nivel junto a las padre. Después todas las funciones de procesado como **get_categories**, **merge_custom_attributes**, etc...se aplican en todo el array de productos/variaciones, y finalmente cuando ya tengo todos los productos y variaciones procesados, es cuando muevo de nivel las variaciones dentro de cada padre. Hacer todo esto en una primera pasada complicaría mucho la lógica y el tiempo total de ejecución apenas varía.

Creo que es la forma más sencilla de hacer todo el procesado, ya que si de primeras meto las variaciones dentro de los padres se complica muchísimo el procesado posterior de todas las funciones.

Reallmente así es como también lo esta haciendo doofeeds hasta ahora. Primero procesaba todo el array de productos y variaciones y al final hacía la recolocación. 

Si no me he explicado bien podemos hacer un meet el lunes y lo vemos con detalle...